nbsphinx_link/__init__.py

```diff
@@ -11,27 +11,95 @@
     "path": "relative/path/to/notebook"
 }
 
+Optionally the "extra-media" key can be added, if you notebook includes
+any media, i.e. images. The value can be an array or a string, with
+path/-s to the media file/-s or directory/-s.
+
 Further keys might be added in the future.
 """
 
-
+import errno
 import json
 import os
+import shutil
 
 from docutils import io, nodes, utils
 from docutils.utils.error_reporting import SafeString, ErrorString
 from nbsphinx import NotebookParser, NotebookError, _ipynbversion
 import nbformat
+from sphinx.util.logging import getLogger
 
 from ._version import __version__
 
 
+def copy_directory(src, dest):
+    """
+    Copies a directory or file from the path ``src`` to ``dest``.
+    Originally from: https://www.pythoncentral.io/how-to-recursively-copy-a-directory-folder-in-python/
+
+    Parameters
+    ----------
+    src : str
+        Path to the source directory or file
+    dest : str
+        Path to the destination directory or file
+    """
+    try:
+        # ensure that files will be overwritten
+        shutil.rmtree(dest, ignore_errors=True)
+        shutil.copytree(src, dest)
+    except OSError as e:
+        # If the error was caused because the source wasn't a directory
+        if e.errno == errno.ENOTDIR:
+            shutil.copy(src, dest)
+        else:
+            print('Directory not copied. Error: {}'.format(e))
+
+
+def collect_extra_media(extra_media, source_file, nb_path):
+    """
+    Collects extra media defined in the .nblink file, with the key 'extra-media'.
+    The extra media (i.e. images) need to be copied on order nbsphinx to properly
+    render the notebooks, since nbsphinx assumes that the files are relative to
+    the .nblink.
+
+    Parameters
+    ----------
+    extra_media : str, list
+        Path/-s to directorie-/s and/or file-/s with extra media.
+    source_file : str
+        Path to the .nblink file.
+    nb_path : str
+        Path to the notebook defined in the .nblink file , with the key 'path'.
+    """
+    source_dir = os.path.dirname(source_file)
+    if isinstance(extra_media, str):
+        extra_media = [extra_media]
+    for extract_media_path in extra_media:
+        if os.path.isabs(extract_media_path):
+            src_path = extract_media_path
+        else:
+            extract_media_relpath = os.path.join(source_dir, extract_media_path)
+            src_path = os.path.normpath(os.path.join(source_dir, extract_media_relpath))
+        target_path = utils.relative_path(nb_path, src_path)
+        target_path = os.path.normpath(os.path.join(source_dir, target_path))
+        if os.path.exists(src_path):
+            copy_directory(src_path, target_path)
```

Comment:
This `copy_directory` call will blow out the contents of `target_path`, which I don't think is safe. I think this can particularly be an issue if the extra-media paths point to a folder above the notebook folder (e.g. `../../data/cats.jpg`). I think we should:
- Ensure that the destination path does is not outside the docs source dir.
- Individually copy each file (overwriting if necessary), rather than an indiscriminate directory deletion.

Reply:
This should only happen, if the ``.nblink`` file would be outside of the docs folder, in which case it won't be discovered by sphinx. So I think we are good.

Reply:
I could also rewrite ``copy_directory`` to just use ``shutil.copy`` and ``os.walk`` if the ``src`` is a directory, which would also save the hustle with license linking of pythoncentral.

Reply:
> I could also rewrite copy_directory to just use shutil.copy, os.makedirs and os.walk if the src is a directory, which would also save the hustle with license linking of pythoncentral.

I think this is the right solution 👍